ntbs-service/Startup.cs

```diff
@@ -472,6 +478,21 @@ private void AddAdImportService(IServiceCollection services, AzureAdOptions azur
 
         }
 
+        private void AddApplicationInsightsMonitoring(IServiceCollection services,
+            ApplicationInsightsOptions applicationInsightsOptions)
+        {
+            services.AddApplicationInsightsTelemetry(new ApplicationInsightsServiceOptions
+            {
+                ConnectionString = Configuration.GetConnectionString(Constants.ConnectionStringApplicationInsights)
+            });
+
+            services.ConfigureTelemetryModule<DependencyTrackingTelemetryModule>((module, o) =>
+            {
+                module.EnableSqlCommandTextInstrumentation =
+                    applicationInsightsOptions.EnableSqlCommandTextInstrumentation ?? false;
+            });
+        }
```

Comment:
So I had hoped to add these insights conditionally (ie if `ApplicationInsights.Enabled != true` then don't add them) but this didn't play nice with the DI into the Razor pages. Having had a look around the accepted solution to this appears to be to give it a `null` connection string (so AI can't talk to the outside world), and then set `telemetryConfiguration.DisableTelemetry` (see below) to get a performance uplift, by not actually collecting any data. It's not perfect but I think it works well enough.

Reply:
Hmm... I'd also prefer to add this conditionally. Could you add some links here of what you found? Is there no way to also conditionally inject the javascript in the Razor pages?

Reply:
So I've had another go, and have pushed an example of how I hoped I could conditionally add the service. Unfortunately, Razor doesn't seem to support conditional dependency injection, because I get the following exception when `ApplicationInsights.Enabled = false`
```
InvalidOperationException: No service for type 'Microsoft.ApplicationInsights.AspNetCore.JavaScriptSnippet' has been registered.
```
yet the call to inject that service is inside of a conditional, but it doesn't seem to help. If you have any ideas I'm all ears

The official Microsoft docs suggest [this](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core#disable-telemetry-dynamically) to disable the insights on the fly. [This StackOverflow](https://stackoverflow.com/questions/32057441/disable-application-insights-in-debug/32074015) post has a discussion about this stuff, but the terminology is a bit different because it is so old. I'm going to go at doing what [this answer](https://stackoverflow.com/a/49458076) suggests to see if that will make it work

Reply:
Ok so that last SO post was the right thing to be reading finally, we can get it to work by adding a class that pretends to be the Application Insights JavaScript to keep the DI happy. It now runs as expected, but I'll let you judge which solution (the one from the first commit of this branch, or the most recent one) is better

Reply:
Thanks for the thorough response! I do think the second way is best - semantically we're saying something like "when app insights is off, we don't want any javascript" whereas the first way is more like "when we don't want app insights, turn it on but pretend it's off".

Reply:
Yeah on second thoughts it does seem like the better approach. All tidied up!